autode/opt/optimisers/trm.py

```diff
@@ -0,0 +1,461 @@
+            self.alpha = max(set_trust, self._min_alpha)
+        elif 0.25 < ratio < 0.75:
+            pass
+        elif 0.75 < ratio < 1.25:
+            # if step taken is within 5% of the trust radius
+            if abs(cart_step_size - self.alpha) / self.alpha < 0.05:
+                self.alpha = min(self.alpha * 1.414, self._max_alpha)
+        elif 1.25 < ratio < 1.5:
+            pass
+        else:  # ratio > 1.5
+            set_trust = 0.5 * min(self.alpha, cart_step_size)
+            self.alpha = max(set_trust, self._min_alpha)
+
+        if (ratio < -1.0) or (ratio > 2.0):
+            # if ratio too low or too high, discard the last point
+            logger.warning(
+                "Energy increased/decreased by large amount,"
+                "rejecting last geometry step"
+            )
+            # copy the previous geometry into current one
+            self._history[-1] = self._history[-2].copy()
+
+        logger.info(
+            f"Ratio = {ratio:.3f}, Current trust radius = {self.alpha:.3f} Å"
+        )
+
+        return None
+
+    def _damp_if_required(self) -> bool:
+        """
+        If the energy and gradient norm are oscillating in the last three
+        iterations, then interpolate between the last two coordinates to
+        damp the oscillation (must skip the quasi-NR step afterwards)
+
+        Returns:
+            (bool): True if damped, False otherwise
+        """
+        # if user does not want, no damping
+        if not self._should_damp:
+            return False
+
+        # allow the optimiser 3 free iterations before damping again
+        if self.iteration - self._last_damp_iter < 3:
+            return False
+
+        # get last three coordinates
+        coords_0, coords_1, coords_2, coords_3 = self._history[-4:]
+
+        # energy changes in last three iters
+        e_change_0_1 = coords_1.e - coords_0.e
+        e_change_1_2 = coords_2.e - coords_1.e
+        e_change_2_3 = coords_3.e - coords_2.e
+
+        is_e_oscillating = (
+            e_change_0_1 * e_change_1_2 < 0.0
+        ) and (  # different sign
+            e_change_1_2 * e_change_2_3 < 0.0
+        )  # different sign
+
+        # the sign of gradient norm change must also flip
+        g_change_0_1 = np.linalg.norm(coords_1.to("cart").g) - np.linalg.norm(
+            coords_0.to("cart").g
+        )
+        g_change_1_2 = np.linalg.norm(coords_2.to("cart").g) - np.linalg.norm(
+            coords_1.to("cart").g
+        )
+        g_change_2_3 = np.linalg.norm(coords_3.to("cart").g) - np.linalg.norm(
+            coords_2.to("cart").g
+        )
+
+        is_g_oscillating = (
+            g_change_0_1 * g_change_1_2 < 0.0
+        ) and (  # different sign
+            g_change_1_2 * g_change_2_3 < 0.0
+        )
+
+        if is_e_oscillating and is_g_oscillating:
+            logger.info("Oscillation detected in optimiser, damping")
+
+            # is halfway interpolation good?
```

Comment:
my instinct would be to average over the oscillating region. would be an interesting bit of science to do!